### ISSUE TITLE
[BUG] User can do upgrade to old version

### ISSUE THREAD

**Issue by lanfon72** [harvester/harvester#5494]
### Describe the bug
The upgrade will started and fail if user use an older ISO to do the upgrade.
![image](https://github.com/harvester/harvester/assets/5169694/6d50c0fa-3e5c-40f5-a902-58a77c35fb2c)
![image](https://github.com/harvester/harvester/assets/5169694/37123a45-7bb9-4640-bb50-4e220900240f)

### To Reproduce
1. Install Harvester with any nodes
2. Create upgrade version
3. Perform upgrade

### Expected behavior
Upgrade should not start

## Environment:
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): **Baremetal DL360G9 3 nodes**
 - Harvester ISO version: **v1.3.0**
 - **ui-source** Option: **Auto**

### Additional context
discovered when implementing test automation of #2431

**Comment by starbops** [comment]
As we're targeting the v1.4 release, the test plan would be...

## Test Plan

1. Prepare a single-node cluster of version v1.4-head (should've contained this fix)
2. Fake the version as it's `v1.4.0` by adding `HARVESTER_SERVER_VERSION` env to the `harvester` deployment
   ```
         ...
         containers:
         - env:
           - name: HARVESTER_SERVER_VERSION
             value: v1.4.0
             ...
   ```
3. Upgrade to v1.3.2
4. The upgrade should be blocked because it's effectively a downgrade behavior

**Comment by lanfon72** [comment]
Tested this bug not be fixed yet, move back to implement.

![image](https://github.com/user-attachments/assets/aa822782-366b-4d6a-9d49-dffb363e1160)
![image](https://github.com/user-attachments/assets/aaecc85b-cd5e-4745-89e5-d6c8a5a33272)


Test Information
----
* Environment: **qemu/KVM 2 nodes and single node**
* Harvester Version: **v1.4-97e60e1-head** (patch to `v1.4.0` by editing *harvester* deployment)
* **ui-source** Option: **Auto**

### Verify Steps
1. Follow **Steps to reproduce** in https://github.com/harvester/harvester/issues/5494#issue-2217383003
    - In `v1.4-head`, we would need to patch *harvester* deployment to `v1.4.0` as https://github.com/harvester/harvester/issues/5494#issuecomment-2375889345 mentioned.

**Comment by starbops** [comment]
Hi @lanfon72, thank you for verifying the fix. I'm trying to reproduce, but no luck. Here's my setup:

1. Install a Harvester v1.4-head (v1.4-ee6361fc-head) 3-node cluster
2. Fake the version to v1.4.0 with environment variable `HARVESTER_SERVER_VERSION` set
3. Kick off the upgrade to v1.3.2
4. When the upgrade repo VM is ready, the upgrade controller immediately detected it's effectively a downgrade behavior and halted the upgrade procedure

<img width="622" alt="image" src="https://github.com/user-attachments/assets/fc07cb4a-39f2-4379-bbb9-748054510ddb">

The Upgrade CR:

```
apiVersion: harvesterhci.io/v1beta1
kind: Upgrade
metadata:
  creationTimestamp: "2024-10-09T03:11:08Z"
  finalizers:
  - wrangler.cattle.io/harvester-upgrade-controller
  generateName: hvst-upgrade-
  generation: 8
  labels:
    harvesterhci.io/latestUpgrade: "true"
    harvesterhci.io/upgradeState: Failed
  name: hvst-upgrade-8ct9t
  namespace: harvester-system
  resourceVersion: "36691"
  uid: d7beb371-127a-463e-b1f1-620ff1f14635
spec:
  image: ""
  logEnabled: true
  version: v1.3.2
status:
  conditions:
  - lastUpdateTime: "2024-10-09T03:17:35Z"
    reason: downgrading is prohibited
    status: "False"
    type: Completed
  - lastUpdateTime: "2024-10-09T03:11:48Z"
    status: "True"
    type: LogReady
  - lastUpdateTime: "2024-10-09T03:12:47Z"
    status: "True"
    type: ImageReady
  - lastUpdateTime: "2024-10-09T03:17:25Z"
    status: "True"
    type: RepoReady
  imageID: harvester-system/hvst-upgrade-8ct9t
  previousVersion: v1.4.0
  repoInfo: |
    release:
      harvester: v1.3.2
      harvesterChart: 1.3.2
      os: Harvester v1.3.2
      kubernetes: v1.28.12+rke2r1
      rancher: v2.8.5
      monitoringChart: 103.0.3+up45.31.1
      minUpgradableVersion: v1.3.1
  upgradeLog: hvst-upgrade-8ct9t-upgradelog
```

The full log messages produced by the `harvester` pod during the upgrade:

```
time="2024-10-09T03:11:08Z" level=info msg="Initialize upgrade harvester-system/hvst-upgrade-8ct9t"
time="2024-10-09T03:11:08Z" level=info msg="Skip store VM state" name=hvst-upgrade-8ct9t namespace=harvester-system
time="2024-10-09T03:11:08Z" level=info msg="Enabling upgrade observability"
time="2024-10-09T03:11:08Z" level=info msg="Initialize UpgradeLog"
time="2024-10-09T03:11:08Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:PreparingLoggingInfra]"
time="2024-10-09T03:11:08Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:PreparingLoggingInfra]"
time="2024-10-09T03:11:08Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:PreparingLoggingInfra]"
time="2024-10-09T03:11:08Z" level=info msg="Check if there are any existing logging-operator"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging Addon is not enabled"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging ManagedChart is not installed"
time="2024-10-09T03:11:08Z" level=info msg="Deploy logging-operator"
time="2024-10-09T03:11:08Z" level=info msg="Check if there are any existing logging-operator"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging Addon is not enabled"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging ManagedChart is not installed"
time="2024-10-09T03:11:08Z" level=info msg="Deploy logging-operator"
time="2024-10-09T03:11:08Z" level=info msg="Check if there are any existing logging-operator"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging Addon is not enabled"
time="2024-10-09T03:11:08Z" level=info msg="rancher-logging ManagedChart is not installed"
time="2024-10-09T03:11:08Z" level=info msg="Deploy logging-operator"
time="2024-10-09T03:11:19Z" level=info msg="Start to create the logging infrastructure for the upgrade procedure"
time="2024-10-09T03:11:19Z" level=info msg="Check if the logging infrastructure is ready"
time="2024-10-09T03:11:20Z" level=info msg="updating log-archive pvc name annotation" kind=UpgradeLog name=hvst-upgrade-8ct9t-upgradelog namespace=harvester-system
time="2024-10-09T03:11:20Z" level=info msg="updating ownerReference" kind=PersistentVolumeClaim name=hvst-upgrade-8ct9t-upgradelog-infra-log-archive-hvst-upgrade-8ct9t-upgradelog-infra-fluentd-0 namespace=harvester-system
time="2024-10-09T03:11:20Z" level=info msg="Check if the logging infrastructure is ready"
time="2024-10-09T03:11:31Z" level=info msg="Check if the logging infrastructure is ready"
time="2024-10-09T03:11:48Z" level=info msg="Check if the logging infrastructure is ready"
time="2024-10-09T03:11:48Z" level=info msg="Logging infrastructure is ready"
time="2024-10-09T03:11:48Z" level=info msg="Check if the log-collecting rules are installed"
time="2024-10-09T03:11:48Z" level=info msg="Start to create the ClusterFlow and ClusterOutput resources for collecting upgrade logs"
time="2024-10-09T03:11:48Z" level=info msg="Check if the log-collecting rules are installed"
time="2024-10-09T03:11:48Z" level=info msg="Start to create the ClusterFlow and ClusterOutput resources for collecting upgrade logs"
time="2024-10-09T03:11:48Z" level=info msg="Check if the log-collecting rules are installed"
time="2024-10-09T03:11:48Z" level=info msg="Log-collecting rules exist and are activated"
time="2024-10-09T03:11:48Z" level=info msg="Logging infrastructure is ready, proceed the upgrade procedure"
time="2024-10-09T03:11:48Z" level=info msg="Creating upgrade repo image"
time="2024-10-09T03:11:48Z" level=info msg="Spin up downloader"
time="2024-10-09T03:11:49Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:LoggingInfraPrepared]"
W1009 03:11:49.354621       7 warnings.go:70] metadata.finalizers: "longhorn.io": prefer a domain-qualified finalizer name to avoid accidental conflicts with other finalizer writers
time="2024-10-09T03:12:47Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:LoggingInfraPrepared]"
time="2024-10-09T03:12:47Z" level=info msg="Starting upgrade repo VM"
time="2024-10-09T03:12:48Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:PreparingRepo]"
time="2024-10-09T03:14:31Z" level=error msg="error syncing 'harvester-system/upgrade-repo-hvst-upgrade-8ct9t': handler VMController.SetDefaultManagementNetworkMacAddress: error fetch controller revision object for vmi upgrade-repo-hvst-upgrade-8ct9t-harvester-system: controllerrevisions.apps \"revision-start-vm-680ca389-0ecb-498d-88be-0ea099a441d9-1\" not found, requeuing"
W1009 03:16:48.790099       7 warnings.go:70] kubevirt.io/v1 VirtualMachineInstancePresets is now deprecated and will be removed in v2.
time="2024-10-09T03:17:25Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:RepoPrepared]"
time="2024-10-09T03:17:25Z" level=warning msg="Repo info retrieval failed with: Get \"http://upgrade-repo-hvst-upgrade-8ct9t.harvester-system/harvester-iso/harvester-release.yaml\": dial tcp 10.53.243.92:80: connect: connection refused"
time="2024-10-09T03:17:35Z" level=info msg="upgrade eligibility check" currentVersion=v1.4.0 minUpgradableVersion="<nil>" name=hvst-upgrade-8ct9t namespace=harvester-system upgradeVersion=v1.3.2
time="2024-10-09T03:17:35Z" level=info msg="handle upgrade harvester-system/hvst-upgrade-8ct9t with labels map[harvesterhci.io/latestUpgrade:true harvesterhci.io/upgradeState:Failed]"
time="2024-10-09T03:17:35Z" level=info msg="Skip restore VM" name=hvst-upgrade-8ct9t namespace=harvester-system
time="2024-10-09T03:17:35Z" level=info msg="Add OwnerReference upgrade-repo-hvst-upgrade-8ct9t-disk-0 to upgrade repo image harvester-system/hvst-upgrade-8ct9t"
time="2024-10-09T03:17:35Z" level=info msg="Delete upgrade repo VM harvester-system/upgrade-repo-hvst-upgrade-8ct9t"
time="2024-10-09T03:17:35Z" level=info msg="Reset RKEConfig and set provisionGeneration to 0"
time="2024-10-09T03:17:35Z" level=warning msg="no original replica-replenishment-wait-interval value set"
time="2024-10-09T03:17:35Z" level=info msg="Update upgradeLog harvester-system/hvst-upgrade-8ct9t-upgradelog"
time="2024-10-09T03:17:35Z" level=info msg="Stop collecting logs"
time="2024-10-09T03:17:35Z" level=info msg="Tearing down the logging infrastructure for upgrade procedure"
```

---

I know it's been a while, but do you happen to have the support bundle for the failure cluster? I checked the version you used to install the cluster based on commit `97e60e1`, which already had the fix included. So I'm unsure why it didn't block the upgrade.

---

I can't reproduce it with single-node or 2-node clusters either (both with v1.4-ee6361fc-head):

![image](https://github.com/user-attachments/assets/841f6e6b-a373-4ed8-81bc-578e29227ef5)

![image](https://github.com/user-attachments/assets/608d88da-be37-4ca6-a289-81c037920c2a)

**Comment by lanfon72** [comment]
Tested on `v1.4.0-rc1` (without update `HARVESTER_SERVER_VERSION`)
![image](https://github.com/user-attachments/assets/3f888b44-6bc0-419b-9883-44cfb9895fa2)

on `v1.4.0-rc2` (without update `HARVESTER_SERVER_VERSION`)
![image](https://github.com/user-attachments/assets/7a6bb2d2-728e-4cbf-8ff0-b826c58bb17e)


The way to create upgrade:
```bash
cat > ./version.yaml << EOF
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: local-upgrade
  namespace: harvester-system
spec:
  isoChecksum: 2b99b0870d3363b393bb7900c53b8a1f529e150519558cea672e9ff8eab19ba0dac3f35688e49cdc2ac3ba0d253017fd3d8790cb229856735e30c524f56f22a7
  isoURL: http://192.170.0.254/harvester-amd64.iso
EOF

kubectl create -f version.yaml
```

I guess that the reason might be `metadata.name` is not a version? (we don't have another way to check ISO's version?)

**Comment by starbops** [comment]
I could still not reproduce the failure case with v1.4.0-rc1 upgrading to v1.3.2 (I didn't update the `HARVESTER_SERVER_VERSION` env).

<img width="1437" alt="image" src="https://github.com/user-attachments/assets/a0a0394c-6dfc-425c-96d0-2a585744e4b3">

Could you share the ISO image `http://192.170.0.254/harvester-amd64.iso` to me? Thank you.

---

> I guess that the reason might be metadata.name is not a version?

We don't reference the `.metadata.name` field in Version or Upgrade CRs. The `.spec.version` and `.status.previousVersion` fields in the Upgrade CR matter.

**Comment by starbops** [comment]
Update: The version CR's name matters. It's propagated to the `.spec.version` of the Upgrade CR. I'll adapt to the `.status.repoInfo.release.harvester` field for our source of truth instead.

---

When you click the upgrade button on the dashboard, it will use the name of the Version CR for the `.spec.version` field of the to-be-created Upgrade CR.
https://github.com/harvester/dashboard/blob/e4342aeba3913a9148d921912a0f2cf7283fce8c/pkg/harvester/components/HarvesterUpgrade.vue#L104 (Thanks to @a110605 for pointing this out)

**Comment by bk201** [comment]
@starbops we need to backport this too.

**Comment by starbops** [comment]
@bk201 That means we also need to backport #5616 to v1.3 too. Since #6767 is based on #5616. Or I'll just find another way to make it work in v1.3 :P

**Comment by bk201** [comment]
https://github.com/harvester/harvester/pull/5616 is a huge refactor, I'm fine not to backport this since in most case the version is correct from our upgrade responder.

**Comment by lanfon72** [comment]
Verified this bug has been fixed.

![image](https://github.com/user-attachments/assets/b5914041-da55-4364-b822-95a5977ade86)


Test Information
----
* Environment: **qemu/KVM 3 nodes**
* Harvester Version: **v1.4-95024f7c-head** (patched to `v1.4.0-rc3`) -> **v1.3.2**
* **ui-source** Option: **Auto**

### Verify Steps
- Same as https://github.com/harvester/harvester/issues/5494#issuecomment-2381784852
